Firestore/Swift/Source/Codable/CodablePassThroughTypes.swift

```diff
@@ -16,11 +16,9 @@
 
 import Foundation
 import FirebaseFirestore
+import FirebaseSharedSwift
 
-internal func isFirestorePassthroughType<T: Any>(_ value: T) -> Bool {
-  return
-    T.self == GeoPoint.self ||
-    T.self == Timestamp.self ||
-    T.self == FieldValue.self ||
-    T.self == DocumentReference.self
-}
+extension GeoPoint: StructureCodingPassthroughType {}
```

Comment:
FWIW, I think the updated protocol-based implementation is more maintainable than what we had before.

Reply:
Thanks.
The only thing that bothers me is that conforming to a protocol states globally that you are a thing that needs to be left alone during structure encoding and decoding.
In actuality it's probably more likely that it's tied to the specific underlying framework (Firestore) - and thus not really a global opt-in. This is of course why I added the boolean flag to opt-in to this behavior in general.

I think it would be a cleaner solution to be able to have separate passthrough types for separate use cases. Thus going back to something like the previous function instead of conforming to a type.
As closures in Swift can't be generic (or at least you need to know the generic parameter when assigning the closure to a property), then it's not possible to pass a 'passthroughTypes()' function directly to the encoder/decoder, but I think you could create a protocol with a requirement of a generic function and then wrap the function in a struct conforming to the closure.

I haven't tried it, but something like:
```
protocol StructureCodingPassthroughTypeResolver {
   func isPassthroughType<T: Any>(_ t: T) -> Bool
}
struct FirestorePassthroughTypeResolver: StructureCodingPassthroughTypeResolver {
 ...
}
````

Then again: perhaps such a design is overkill since the only use case for the passthrough types is currently Firestore...

What do you think?

Reply:
The function requirement of the protocol could even be static, so that no actual instance of a resolver needs to be passed around, just the type...

Reply:
Do you think we might be able to expand on the id of the `StructureCodingPassthroughTypeResolver` and make it handle both the `@DcoumentId` annotation as well as the Passthrough types? We could call the new `TypeResolver` for every type and can optionally implement custom encoding.

```
protocol TypeResolver {
  func encode<T : Encodable>(_ value: T, forKey key: Key) : Optional<Data>;
  func decode<T : Decodable>(_ type: T.Type, forKey key: Key) : Decode<T>;
}

struct FirestoreTypeResolver: TypeResolver {
  func encode<T : Encodable>(_ value: T, forKey key: Key) : Optional<Data> {
    // Handles DocumenID, GeoPoint, Timestamp, DocumentReference. Returns Optional.none for all other types
   }

  ...
}
```

The DocumentReference can then be passes to FirestoreTypeResolver, which means that we no longer need to add it to `userInfo`.

I am not sure how feasible this is without implementing it first, but you might some more expertise here.

Reply:
I've tried looking at it, and I am not certain that the two concepts can be bridged.

The passthrough type checking is being used in the box/unbox methods of the encoder/decoder and thus don't have access to any keys.

Keys are actually also ignored in the documentid decoding (since the point is that there is no data at the currently parsed level in the input, that corresponds to the id - it comes from somewhere else).

So the signatures should perhaps rather be `encode<T>(_ value: T): Optional<T>` and similar for decode - to allow passing types through. So basically the existing boolean function check converted to an optional of the input...

But for decoding `DocumentID` this type signature is not really a fit since there's no 'DocumentID' _value_ to pass along to the decoding...

---
I did try implementing a passthrough type resolver - and I enhanced discoverability a little by introducing a `public func decode<T: Decodable>(_ t: T.Type, from data: Any, in reference: DocumentReference)`
overload on `Firestore.Decoder` (aka `StructureDecoder`).

I updated the PR branch with the new refactor. Let me know what you think.

Reply:
I just had a different thought:
Instead of Firestore.Decoder being a typealias for StructureDecoder (and ditto encoder), it could be a really thin wrapper, so that the correct configuration is always guaranteed. 
What do you think about that idea?

Reply:
In the latest version of the PR I tried implementing the lightweight wrapper types around the encoder and decoder, and that does indeed give a much nicer API at the use point.